src/value.rs

```diff
@@ -27,33 +29,125 @@ impl NumericValue {
     }
 }
 
-#[derive(PartialEq, Debug)]
+impl PartialOrd for NumericValue {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        match (self.0.as_i64(), self.0.as_u64(), self.0.as_f64()) {
+            (Some(a), _, _) => match (other.0.as_i64(), other.0.as_u64(), other.0.as_f64()) {
+                (Some(b), _, _) => Some(a.cmp(&b)),
+                (_, Some(b), _) => {
+                    if a < 0 {
+                        Some(std::cmp::Ordering::Less)
+                    } else {
+                        (a as u64).partial_cmp(&b)
+                    }
+                }
+                (_, _, Some(b)) => (a as f64).partial_cmp(&b),
+                _ => None,
+            },
+            (None, Some(a), None) => match (other.0.as_i64(), other.0.as_u64(), other.0.as_f64()) {
+                (None, Some(b), None) => Some(a.cmp(&b)),
+                (Some(b), None, None) => {
+                    if b < 0 {
+                        Some(std::cmp::Ordering::Greater)
+                    } else {
+                        a.partial_cmp(&(b as u64))
+                    }
+                }
+                (None, None, Some(b)) => (a as f64).partial_cmp(&b),
+                _ => None,
+            },
+            (None, None, Some(a)) => match (other.0.as_i64(), other.0.as_u64(), other.0.as_f64()) {
+                (None, None, Some(b)) => a.partial_cmp(&b),
+                (Some(b), None, None) => a.partial_cmp(&(b as f64)),
+                (None, Some(b), None) => a.partial_cmp(&(b as f64)),
+                _ => None,
+            },
+            _ => None, // not representable using rust types
+        }
+    }
+}
```

Comment:
Dealing with the underlying json `Number` is quite hard:
- It does not implement Partial Order
- It can represent numbers rust can not represent (using standard types). E.g. Very big numbers, Floating point numbers like 0.5, ...
- It cannot represent some numbers rust can represent. E.g. NaN and Infinity

By wrapping Number `pub struct NumericValue(pub(crate) serde_json::Number);`, we move all those responsibilities very close to the user and need to deal with representation errors as well with conversions very late.

What do you think if we make NumericValue a simple enum:
```rust
pub enum NumericValue{
    UnsignedInteger(u64),
    SignedInteger(i64),
    FloatingPoint(f64),
    NotRepresentable
}
```
and then do all the conversion at the earliest point possible?